packages/order-utils/src/rich_reverts.ts

```diff
@@ -0,0 +1,376 @@
+import { ObjectMap, OrderStatus } from '@0x/types';
+import { AbiEncoder, BigNumber } from '@0x/utils';
+import { DataItem, RichRevertAbi } from 'ethereum-types';
+import * as ethUtil from 'ethereumjs-util';
+import * as _ from 'lodash';
+import { inspect } from 'util';
+
+type ArgTypes = string | BigNumber | number | boolean;
+type ValueMap = ObjectMap<ArgTypes | undefined>;
+type Decoder = (hex: string) => ValueMap;
+
+// tslint:disable:max-classes-per-file custom-no-magic-numbers
+
+export enum FillErrorCodes {
+    InvalidTakerAmount,
+    TakerOverpay,
+    Overfill,
+    InvalidFillPrice,
+}
+
+export enum SignatureErrorCodes {
+    BadSignature,
+    InvalidLength,
+    Unsupported,
+    Illegal,
+    WalletError,
+    ValidatorError,
+}
+
+export enum AssetProxyDispatchErrorCodes {
+    InvalidAssetDataLength,
+    UnknownAssetProxy,
+}
+
+export enum TransactionExecutionErrorCodes {
+    NoReentrancy,
+    AlreadyExecuted,
+    BadSignature,
+    FailedExecution,
+}
+
+const ADDRESS_SIZE = 20;
+
+/**
+ * Base type for rich revert reasons.
+ */
+export abstract class RichRevertReason {
```

Comment:
Does it make sense for this to be in the `order-utils` package? `contract-wrappers` feels more natural to me since we're decoding our contact's reverts specifically.